### ISSUE TITLE
Data_oM: Domain object does not define whether extremes are included or not

### ISSUE THREAD

**Issue by alelom** [BHoM/BHoM#1162]
#### Description:

As title. This is a problem because:
- certain checks may require extremes to be included or not included
- it is not clear what the assumption is - are extremes included or not by default?

## 1) Default value

Currently, methods in the BHoM_Engine like `IsInRange` seem to default to the fact that the Domain does **not** include the extremes.

This may be seen as counter-intuitive e.g. Grasshopper does include the extremes in the domain:
![image](https://user-images.githubusercontent.com/6352844/105172533-af65ec80-5b17-11eb-8e04-94f2088f2f24.png)

## 2) Make the "extremes inclusion" a defining property of the Domain 
Suggestion to add a `DomainExtremesInclusion` enum property to the Domain:

```cs
    public enum DomainExtremesInclusion
    {
        IncludeBoth,
        IncludeMin,
        IncludeMax,
        ExcludeBoth
    }
```

this would remove any confusion around the extremes and give the option on whether to include them or not, and which.

**Comment by IsakNaslundBh** [comment]
Trying to look how we have dealt with this elsewhere, for example on the `BoundingBox`, which is basically a 3D domain. There the `IsContaining` method is accepting a boolean argument if the extremes, or in this case edge, should be accepted or not:

https://github.com/BHoM/BHoM_Engine/blob/028418b935df1359c310159eb5a6b0137e3d5480/Geometry_Engine/Query/IsContaining.cs#L37-L56

Same goes for the curve containment methods.

That solution would leave the object definition as it is, but include an argument on the methods instead, which would align with the above.

Saying that, can see the case of the `Domain` to be a slightly different case than the geometrical, with it actually making sense to store it on the class itself, being able to differentiate between `(0,1)`, `(0,1]`, `[0,1)` and `[0,1]`, so happy with adding something like an enum to help define that.

An option could be 2 booleans like `StartInclusie` and `EndInclusive` or similar, but not feeling to strongly.

TL;DR: think it is fine to update the class, just wanted to give some context 😄

**Comment by alelom** [comment]
> Saying that, can see the case of the `Domain` to be a slightly different case than the geometrical, with it actually making sense to store it on the class itself,

We're exactly on the same page. The Domain, as opposed from the Box, cannot be fully defined unless we say whether it contains the extremes or not – after all, `(0,1)`, `(0,1]`, `[0,1)` and `[0,1]` are all different Domains! This should be reflected by a property of the Domain.

> An option could be 2 booleans like StartInclusie and EndInclusive or similar, but not feeling to strongly.

My personal preference is an `enum` rather than 2 booleans, it feels less fiddly to me.

**Comment by FraserGreenroyd** [comment]
> > An option could be 2 booleans like StartInclusie and EndInclusive or similar, but not feeling to strongly.
> 
> My personal preference is an `enum` rather than 2 booleans, it feels less fiddly to me.

An enum allows for options to be added without handled extra properties and without immediate changes in the code, but anything wanting to handle it still needs to be written. So it's 6 of 1, half a dozen of the other really... enums seem cleaner, but that's my personal view, but booleans make it more explicit.

Join me next time for more comments that sit on the fence :wink:
